### DISCUSSION TITLE
High CPU usage

### DISCUSSION THREAD

**Discussion by a-khakimov** [savonet/liquidsoap#3566] [User Support]
I want to start a YouTube stream with the following configuration. I have a playlist of several songs and a background picture.

```liq
set("log.file.path","/tmp/liquidsoap.log")
set("log.level", 3)
audio.samplerate.set(44100)
video.frame.width.set(1920)
video.frame.height.set(1080)
video.frame.rate.set(25)

video_source = single("/path/to/imj.jpg")
audio = mksafe(playlist(mode='randomize', reload=1, reload_mode="rounds", '/path/to/playlist.m3u'))

stream = mux_video(video=video_source, audio)

full = ffmpeg.encode.audio_video(
    %ffmpeg(
        %audio(codec="aac", b="128k"),
        %video(codec="libx264", b="1000k")
    ),
    stream
)

flv = %ffmpeg(
    format="flv",
    %audio.copy,
    %video.copy
)

flv_stream = ffmpeg.filter.bitstream.extract_extradata(full)

output.youtube.live.rtmp(key=youtube_key, encoder=flv, flv_stream)
```

With this configuration, I get too high CPU usage

<img width="855" alt="Screenshot 2023-11-27 222825" src="https://github.com/savonet/liquidsoap/assets/33376759/735668f7-f453-481e-b010-26feaa775a71">

The question is whether such a CPU load is normal (I think not). If not, what is wrong with my configuration?

```bash
$ liquidsoap --version
Liquidsoap 2.1.4
Copyright (c) 2003-2022 Savonet team
Liquidsoap is open-source software, released under GNU General Public License.
See <http://liquidsoap.info> for more information.
```

**Comment by vitoyucepi** [comment]
Hi @a-khakimov,
Could you please specify the CPU model of your computer?
I think the main load comes from video encoding. The default libx264 preset is *medium*, you could change it to something like *superfast* or *ultrafast*.
Have a look at [page 151 of the book](https://www.liquidsoap.info/book/book.pdf).

**Comment by a-khakimov** [comment]
I want to run a stream on a rented server with such a CPU

```bash
$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 15
model           : 6
model name      : Common KVM processor
stepping        : 1
microcode       : 0x1
cpu MHz         : 2197.450
cache size      : 16384 KB
physical id     : 0
siblings        : 2
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault pti ssbd ibrs ibpb fsgsbase bmi1 avx2 smep bmi2 xsaveopt arat
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips        : 4394.90
clflush size    : 64
cache_alignment : 128
address sizes   : 46 bits physical, 48 bits virtual
power management:

processor       : 1
vendor_id       : GenuineIntel
cpu family      : 15
model           : 6
model name      : Common KVM processor
stepping        : 1
microcode       : 0x1
cpu MHz         : 2197.450
cache size      : 16384 KB
physical id     : 0
siblings        : 2
core id         : 1
cpu cores       : 2
apicid          : 1
initial apicid  : 1
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm cpuid_fault pti intel_ppin ssbd ibrs ibpb fsgsbase bmi1 avx2 smep bmi2 xsaveopt arat
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips        : 4394.90
clflush size    : 64
cache_alignment : 128
address sizes   : 46 bits physical, 48 bits virtual
power management:
```

**Comment by a-khakimov** [comment]
Yes, it looks like you're right, I set the superfast parameter and the CPU load decreased by about half

**Comment by a-khakimov** [comment]
Is it possible to stream without video processing? So that there is no load on the processor

**Reply by vitoyucepi** [reply]
You can use `%audio.copy` and `%video.copy` to copy already encoded data, but I don't think this will suit your case.
The alternative solution is to use hardware video encoding, but most virtual machine CPUs don't support it, and allocating a GPU is not very cost effective.

**Reply by gAlleb** [reply]
@vitoyucepi ! Hello! Since I've found you here:)) In order not to start a new thread, could you explain to me the difference between `main = source.mux.video(video=background_video, audio)` and
`main = source(source.tracks(audio).{video=source.tracks(background_video).video})` when applied to simple combining of audio and video background? Any difference at all in such case? Maybe in CPU usage? 

Thanks!

**Reply by vitoyucepi** [reply]
Hi @gAlleb,
I think there's no difference.
https://github.com/savonet/liquidsoap/blob/7d6917edbe77d87e39a5deb82324f6cb068e2dfe/src/libs/tracks.liq#L26-L28
https://github.com/savonet/liquidsoap/blob/7d6917edbe77d87e39a5deb82324f6cb068e2dfe/src/libs/tracks.liq#L12-L14

**Reply by gAlleb** [reply]
Thanks! I thought so)

**Comment by a-khakimov** [comment]
It's probably worth starting a new discussion, but I'll ask here. I'm running a stream with the configuration I applied above. I noticed that after 2-3 days the stream freezes. I thought it was a problem with the Internet or YouTube, but no. I tried to turn on the stream on different devices and the same behavior everywhere. The problem is that the stream freezes. After restarting the process, the stream starts running again.

**Reply by vitoyucepi** [reply]
Freezes with 0% CPU or 100% CPU usage?

**Reply by a-khakimov** [reply]
I specified libx264 preset is ultrafast so the CPU usage is about 50%

```
full = ffmpeg.encode.audio_video(
    %ffmpeg(
        %audio(codec="aac", b="96k"),
        %video(codec="libx264",pixel_format="yuv420p",crf=10,preset="ultrafast")
    ),
    source_mux
)
```

**Reply by vitoyucepi** [reply]
> I specified libx264 preset is ultrafast so the CPU usage is about 50%

And what about freeze? How does the process behave when it freezes?
Does the CPU utilization go up to 100% or down to 0%?

<details>
<summary>Here is how I would do a refactoring of your script</summary>

```ruby
## Maybe log to stdout instead if you're using systemd service?
settings.log.file.path.set("/tmp/liquidsoap.log")

## The default log level is 3
# settings.log.level.set(3)

## The default samplerate is 44100
# settings.audio.samplerate.set(44100)

## The default video resolution is 1280x720
settings.video.width.set(1920)
settings.video.height.set(1080)

## The default framerate is 25
# settings.video.framerate.set(25)

video_source = single("/path/to/imj.jpg")
audio = playlist(mode='randomize', reload=1, reload_mode="rounds", '/path/to/playlist.m3u')
audio = mksafe(audio)

stream = mux_video(video=video_source, audio)

flv_format = %ffmpeg(
  format="flv",
  %audio(codec="aac", b="96k"),
  %video(codec="libx264", pixel_format="yuv420p", b="500k", preset="ultrafast")
)

output.youtube.live.rtmp(key=youtube_key, encoder=flv_format, stream)
```

</details>
